### PR TITLE
#10514 - FeatureEditor filter by geometric area

### LINKED ISSUE #10514: FeatureEditor - Restrict editing by geometric area
> Target branch : master

## Description

> link to https://groups.google.com/g/mapstore-developers/c/6KhmNTJzpDc

This improvement will change the featureEditor plugin to allow to limit attributes table features loading by a geometric area (WKT or GeoJSON) for non ADMIN users.

To activate this restriction, the user (with correct access) will configure FeatureEditor plugin to set a new `restricedArea` params section : 

```
    "restrictedArea": {
      "url": "/my_geojson_or_wkt",
      "raw": "POLYGON ((-1.625007 48.12657, -1.625007 48.134647, -1.603028 48.134647, -1.603028 48.12657, -1.625007 48.12657))",
      "operator": "WITHIN"
    }
```

**Where :**

- `Operator <string>` : WITHIN or INTERSECTS or CONTAINS
- `raw <string>` : geometry GEOJSON or WKT
- `url <string>` : path or URI _(Will be use instead of raw if both are defined)_

About sync and request, this area will be insert in SpatialField array filters with Viewport and others Filters.

In addition, an icon show that the restriction by area is currently active (left) :

image

### PR DESCRIPTION
**Pull request by Gaetanbrl**
## Description

This improvement will change the featureEditor plugin to allow to limit attributes table features loading by a geometric area (WKT or GeoJSON) for non ADMIN users.

This improvement don't use GeoFence and use a simple geometry.

>https://groups.google.com/g/mapstore-developers/c/6KhmNTJzpDc/m/12J-2t9BAAAJ

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [ ] Bugfix
 - [x] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [ ] Other... Please describe:

## Issue

**What is the current behavior?**

User can filter attribute table by custom filter or viewport.

#10514 

**What is the new behavior?**

If configured, attribute table keep natives filters and use an additional geometric filter that can be create from URL (e.g geOrchestra restricted area) or from RAW (GEOJSON or WKT).

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

**Comment by tdipisa** [comment]
Hi @Gaetanbrl thank you for contributing. We will review asap.

File: web/client/selectors/featuregrid.js [hunk 1]
```diff
@@ -202,36 +202,87 @@ export const isEditingAllowedSelector = (state) => {
...
             ]
         } : {};
     }
 );
+
+/**
+ * Create spatialField filters array.
+ * Contains filters from viewportFilter, restrictedArea, exsting WFS filter
+ */
+export const additionnalGridFilters = (state) => {
```
Comment (offtherailz) [inline web/client/selectors/featuregrid.js:284]: Using this approach makes the sync with map not working. While it still makes sense with viewport filter (because the map already shows the features in map), here this makes things a little unconsistent, and it could be practical for the user to see only the allowed rows on the map. What do you think?

A more functional approach is to add / remove the filter to the `query` object, as well as the other filters, like quick filter, do. 
image

The  `filters` [section](https://docs.mapstore.geosolutionsgroup.com/en/v2024.02.00/developer-guide/LayerFilter/)  recently added to the filter may help you to manage it in a easier way.

Basically you can add to the filter a section filters: that can contain and use cql filter to apply some additional logic. 
You can identify these filters by id, adding your own property, to manage them (e.g. adding/removing).

File: web/client/plugins/FeatureEditor.jsx [hunk 2]
```diff
@@ -79,6 +79,9 @@ import {isViewportFilterActive} from "../selectors/featuregrid";
   * @prop {array} cfg.showFilterByViewportTool Show button to toggle filter by viewport in toolbar.
   * @prop {object} cfg.dateFormats Allows to specify custom date formats ( in [ISO_8601](https://en.wikipedia.org/wiki/ISO_8601)  format) to use to display dates in the table. `date` `date-time` and `time` are the supported entries for the date format. Example:
   * @prop {boolean} cfg.showPopoverSync default false. Hide the popup of map sync if false, shows the popup of map sync if true
+  * @prop {string} cfg.restrictedArea.url Geometry definition as WKT or GeoJSON loaded from URL or path.
+  * @prop {string} cfg.restrictedArea.raw Geometry definition as WKT or GeoJSON.
+  * @prop {string} cfg.restrictedArea.operator Spatial operation to performed between features and the given geometry.
```
Comment (offtherailz) [inline web/client/plugins/FeatureEditor.jsx:84]:
```suggestion
  * @prop {object} cfg.restrictedArea object containing settings for restricted area. If present, it restricts the editing area to the given geometry. It requires at least `url` or `raw` to be defined.
  * @prop {string} cfg.restrictedArea.raw Geometry definition as WKT or GeoJSON. This attribute allows to define the geometry directly in the configuration.
  * @prop {string} cfg.restrictedArea.url Geometry definition as WKT or GeoJSON loaded from URL or path. If present, this wins over the raw geometry configuration. By default, the filter will use `CONTAINS` if not defined.
```

**Review by offtherailz — CHANGES_REQUESTED** [review]
Thank you for the PR.  In this review I reported all the things that I found that may look unconsistent or problematic.

- [ ] see my comments inline
- [ ] The original proposal and issue was regarding to restrict editing, but in this implementation, also the view mode encounter this filter. Also this makes the things a little unclear. Maybe the filter should be applied only in edit mode?
- [ ] I found a bug that I didn't investigated, so I write about it here, instead of inline. With this tool active, I see an error of "max stack exceed" as well as I try to apply a quick filter.
    https://github.com/user-attachments/assets/0193f17b-bfc2-41a2-a152-184e9482c445
    I didn't debug so I don't know the exact reason. Anyway using `spatialField` can problematic, because too related to the advanced filter form . I'd suggest to use "[filters](https://docs.mapstore.geosolutionsgroup.com/en/v2024.02.00/developer-guide/LayerFilter/)" instead, as reported in another inline comment.

**Comment by Gaetanbrl** [comment]
@offtherailz Thank you for this detailed review.
I'll think about functional proposals with and try to plan technical changes.

**Comment by Gaetanbrl** [comment]
The source branch is now rebase / align with master (https://github.com/geosolutions-it/MapStore2/commit/8e2d4433427ce5b7cf7628c8d51f724f29d2bf18)

File: web/client/epics/featuregrid.js [hunk 3]
```diff
@@ -1283,3 +1298,34 @@ export const resetViewportFilter = (action$, store) =>
         return viewportFilter(store.getState()) !== null ? Rx.Observable.of(setViewportFilter(null))
             : Rx.Observable.empty();
     });
+
+export const requestRestrictedArea = (action$, store) =>
+    action$.ofType(OPEN_FEATURE_GRID, LOGIN_SUCCESS)
+        .filter(() => {
+            return !isAdminUserSelector(store.getState())
```
[the remark this answers is not in the document]
Comment (Gaetanbrl) [reply web/client/epics/featuregrid.js:1305]: Indeed, we have think avout authenticated users (to works with geOrchestra restricted org area) first and forget anonymous users.

To apply this restriction, we needs to confirm if restricted area is apply in edit mode only or not as explain in global review comment.

[the remark this answers is not in the document]
Comment (Gaetanbrl) [reply web/client/epics/featuregrid.js:1305]: > I think we should at least apply the same restrictedArea to the anonymous users?

After discussion, we will also apply the restriction to anonymous users to be compliant with MapStore2 behavior.

File: web/client/components/data/featuregrid/toolbars/Toolbar.jsx [hunk 4]
```diff
@@ -35,7 +35,23 @@ const standardButtons = {
...
+            bsStyle="warning"
+            tooltipId="featuregrid.toolbar.restrictedByArea"
+            disabled
+            style={isEmpty(restrictedArea) ? {
+                width: 0,
+                padding: 0,
+                borderWidth: 0
+            } : {}}>
+            <Glyphicon glyph="1-point-dashed" />
+        </Button>);
```
[the remark this answers is not in the document]
Comment (Gaetanbrl) [reply web/client/components/data/featuregrid/toolbars/Toolbar.jsx:53]: Thanks. This change will be made.

File: web/client/selectors/featuregrid.js (hunk 1 above)
Comment (Gaetanbrl) [reply web/client/selectors/featuregrid.js:284]: > it could be practical for the user to see only the allowed rows on the map. What do you think?

Exactly, this is the expected behavior

> Using this approach makes the sync with map not working. 

Not really sur to understand. additionnalGridFilters methods is not dependent on the screen filter but but also uses it if activate. Maybe my code is wrong with I wanted to do.

> Basically you can add to the filter a section filters: that can contain and use cql filter to apply some additional logic.

I will change the code to use filters section as you expect.

Thanks for this proposals @offtherailz .

**Comment by Gaetanbrl** [comment]
> Thank you for the PR. In this review I reported all the things that I found that may look unconsistent or problematic.
> 
> * [ ]  see my comments inline
> * [ ]  The original proposal and issue was regarding to restrict editing, but in this implementation, also the view mode encounter this filter. Also this makes the things a little unclear. Maybe the filter should be applied only in edit mode?
> * [ ]  I found a bug that I didn't investigated, so I write about it here, instead of inline. With this tool active, I see an error of "max stack exceed" as well as I try to apply a quick filter.
>   https://github.com/user-attachments/assets/0193f17b-bfc2-41a2-a152-184e9482c445
>   I didn't debug so I don't know the exact reason. Anyway using `spatialField` can problematic, because too related to the advanced filter form . I'd suggest to use "[filters](https://docs.mapstore.geosolutionsgroup.com/en/v2024.02.00/developer-guide/LayerFilter/)" instead, as reported in another inline comment.

After some discussions, we will make the expected changes (see comments).

> Maybe the filter should be applied only in edit mode?

About this point, the natif request is to limit view by an area in edit mode and read mode.

The natif use case is to limit view and edition according to geOrchestra restricted area (this area is applied to the user's organization via geOrchestra console app).

So, we propose to add a configuration to manage 2 options :

- allow to restrict edition only (to take your request into account)
- allow to restrict edition and view


> I see an error of "max stack exceed" as well as I try to apply a quick filter.

Needs investigation. I've seen this error before with CQL filters and geoServer limits (cross layer filter limit ?).

File: web/client/selectors/featuregrid.js [hunk 5]
```diff
@@ -202,36 +202,87 @@ export const isEditingAllowedSelector = (state) => {
...
+    spatialFieldFilters,
     describeSelector,
     isFilterByViewportSupported,
     (viewportFilterIsActive, box, projection, spatialField = [], describeLayer, viewportFilterIsSupported) => {
         const attribute = findGeometryProperty(describeLayer)?.name;
         const existingFilter = spatialField?.operation ? [spatialField] : spatialField;
         return viewportFilterIsActive && viewportFilterIsSupported ? {
             spatialField: [
-                ...existingFilter,
+                // avoid restricted area filter dupplication
```
Comment (MV88) [inline web/client/selectors/featuregrid.js:231]:
```suggestion
                // avoid restricted area filter duplication
```

File: web/client/selectors/featuregrid.js [hunk 6]
```diff
@@ -202,36 +202,87 @@ export const isEditingAllowedSelector = (state) => {
...
+    }
+);
+
+export const restrictedAreaFilter = createShallowSelectorCreator(isEqual)(
+    restrictedAreaSelector,
+    spatialFieldFilters,
+    viewportFilter,
+    projectionSelector,
+    describeSelector,
+    state => restrictedAreaOperatorSelector(state),
```
Comment (MV88) [inline web/client/selectors/featuregrid.js:254]:
```suggestion
   restrictedAreaOperatorSelector,
```

File: web/client/plugins/FeatureEditor.jsx [hunk 7]
```diff
@@ -175,7 +183,9 @@ const EditorPlugin = connect(
                 virtualScroll: this.props.virtualScroll ?? true,
                 editingAllowedRoles: this.props.editingAllowedRoles,
                 editingAllowedGroups: this.props.editingAllowedGroups,
-                maxStoredPages: this.props.maxStoredPages
+                maxStoredPages: this.props.maxStoredPages,
+                restrictedAreaUrl: this.props.restrictedAreaUrl,
```
Comment (MV88) [inline web/client/plugins/FeatureEditor.jsx:187]: this is not present in config, only restrictedArea with inside three params:
- raw
- operator
- url

File: web/client/reducers/featuregrid.js [hunk 8]
```diff
@@ -156,7 +157,9 @@ function featuregrid(state = emptyResultsState, action) {
             editingAllowedRoles: action.options.editingAllowedRoles || state.editingAllowedRoles || ["ADMIN"],
             editingAllowedGroups: action.options.editingAllowedGroups || state.editingAllowedGroups || [],
             virtualScroll: !!action.options.virtualScroll,
-            maxStoredPages: action.options.maxStoredPages || 5
+            maxStoredPages: action.options.maxStoredPages || 5,
+            restrictedAreaUrl: action.options.restrictedAreaUrl || "",
```
Comment (MV88) [inline web/client/reducers/featuregrid.js:161]: this is not needed as url is already inside restrictedArea

File: web/client/reducers/featuregrid.js [hunk 9]
```diff
@@ -441,6 +444,9 @@ function featuregrid(state = emptyResultsState, action) {
     case MAP_CONFIG_LOADED: {
         return {...state, ...get(action, 'config.featureGrid', {})};
     }
+    case SET_RESTRICTED_AREA: {
```
Comment (MV88) [inline web/client/reducers/featuregrid.js:447]: add unit test for this

File: web/client/reducers/featuregrid.js [hunk 10]
```diff
@@ -156,7 +157,9 @@ function featuregrid(state = emptyResultsState, action) {
             editingAllowedRoles: action.options.editingAllowedRoles || state.editingAllowedRoles || ["ADMIN"],
             editingAllowedGroups: action.options.editingAllowedGroups || state.editingAllowedGroups || [],
             virtualScroll: !!action.options.virtualScroll,
-            maxStoredPages: action.options.maxStoredPages || 5
+            maxStoredPages: action.options.maxStoredPages || 5,
+            restrictedAreaUrl: action.options.restrictedAreaUrl || "",
+            restrictedArea: action.options.restrictedArea || {}
```
Comment (MV88) [inline web/client/reducers/featuregrid.js:162]: add unit test for this

File: web/client/selectors/featuregrid.js (hunk 1 above)
Comment (MV88) [reply web/client/selectors/featuregrid.js:284]: is this done?

File: web/client/utils/FeatureGridUtils.js [hunk 11]
```diff
@@ -392,3 +394,38 @@ export const supportsFeatureEditing = (layer) => includes(supportedEditLayerType
  * @returns {boolean} flag
  */
 export const areLayerFeaturesEditable = (layer) =>  !layer?.disableFeaturesEditing && supportsFeatureEditing(layer);
+
+export const isWKT = (wktString) => {
```
[the remark this answers is not in the document]
Comment (MV88) [reply web/client/utils/FeatureGridUtils.js:398]: is this done?

File: web/client/epics/featuregrid.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (MV88) [reply web/client/epics/featuregrid.js:1305]: is this done?

File: web/client/components/data/featuregrid/toolbars/Toolbar.jsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (MV88) [reply web/client/components/data/featuregrid/toolbars/Toolbar.jsx:53]: is this suggestion applied?

File: web/client/plugins/FeatureEditor.jsx (hunk 2 above)
Comment (MV88) [reply web/client/plugins/FeatureEditor.jsx:84]: is this done?

**Review by MV88 — CHANGES_REQUESTED** [review]
Hi @Gaetanbrl I resumed the review recently and i think you are still working on it right?
if so consider this as a partial review, since changes still have to be submitted yet

all of the following must have unit tests covering changes
- selectors 
- reducers 
- actions 
- epics

please checkout also inline comments

also if you can go through the various conversation and add a comment to say which can be resolved or not.

Thanks

File: web/client/utils/FeatureGridUtils.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (Gaetanbrl) [reply web/client/utils/FeatureGridUtils.js:398]: Not yet.

File: web/client/epics/featuregrid.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gaetanbrl) [reply web/client/epics/featuregrid.js:1305]: Not yet.

File: web/client/components/data/featuregrid/toolbars/Toolbar.jsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (Gaetanbrl) [reply web/client/components/data/featuregrid/toolbars/Toolbar.jsx:53]: Not yet.

File: web/client/selectors/featuregrid.js (hunk 1 above)
Comment (Gaetanbrl) [reply web/client/selectors/featuregrid.js:284]: Not yet.

File: web/client/plugins/FeatureEditor.jsx (hunk 2 above)
Comment (Gaetanbrl) [reply web/client/plugins/FeatureEditor.jsx:84]: Not yet.

**Comment by Gaetanbrl** [comment]
@MV88 thank you for this review and your precious time (which we all miss).

> I resumed the review recently and i think you are still working on it right?

Yes, but last comment ([here](https://github.com/geosolutions-it/MapStore2/pull/10515#issuecomment-2500206650)) waiting for some feedback about [this comment](https://github.com/geosolutions-it/MapStore2/pull/10515#issuecomment-2443828702) and [this comment](https://github.com/geosolutions-it/MapStore2/pull/10515#issuecomment-2500203185).
Especially on my counter proposal (see comment).

So, I didn't move forward, having no additional interaction.

I've been away since your last comment, I'll get back now to work on this contribution to change technical point but I always need a feedback about functional elements.

File: web/client/selectors/featuregrid.js [hunk 12]
```diff
@@ -202,36 +202,87 @@ export const isEditingAllowedSelector = (state) => {
     })(state);
     return (canEdit || isAllowed) && !isCesium(state);
 };
+
+export const restrictedAreaSrcSelector = state => get(state, "featuregrid.restrictedArea");
```
[the remark this answers is not in the document]
Comment (Gaetanbrl) [reply web/client/selectors/featuregrid.js:206]: Modified and pushed. Thanks to you.

File: web/client/selectors/featuregrid.js (hunk 5 above)
Comment (Gaetanbrl) [reply web/client/selectors/featuregrid.js]: Modified and pushed. Thanks to you.

File: web/client/plugins/FeatureEditor.jsx (hunk 2 above)
Comment (Gaetanbrl) [reply web/client/plugins/FeatureEditor.jsx]: Modified and pushed. Thanks to you.

File: web/client/components/data/featuregrid/toolbars/Toolbar.jsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (Gaetanbrl) [reply web/client/components/data/featuregrid/toolbars/Toolbar.jsx]: About this change : 

> replace Button by TButton

Modification made and pushed. Thanks.

File: web/client/utils/FeatureGridUtils.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (Gaetanbrl) [reply web/client/utils/FeatureGridUtils.js:412]: After some code seek,
I will use [toGeoJson (from ogc/WKS utils)](https://github.com/geosolutions-it/MapStore2/blob/02520e5c0f3d97fb5df1a24c6b51802e5ba61ceb/web/client/utils/ogc/WKT/toGeoJSON.js#L147). Seems correct to detect if a string return a correct geometry.

File: web/client/epics/featuregrid.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Gaetanbrl) [reply web/client/epics/featuregrid.js:1295]: Via https://github.com/geosolutions-it/MapStore2/pull/10515/commits/be5120d0027b335d7f945ce3706d813f31c78b6c commit , everyone will be restricted by default excepted admins.

@offtherailz is this enough ?

**Comment by Gaetanbrl** [comment]
@MV88 you answer "Is this done ?" about some remaining returns :

- [here](https://github.com/geosolutions-it/MapStore2/pull/10515#discussion_r1925726388)
- [here](https://github.com/geosolutions-it/MapStore2/pull/10515#discussion_r1925723669)

It would be really nice if you could give me some feedback on this

- https://github.com/geosolutions-it/MapStore2/pull/10515#issuecomment-2443828702
- especially about "How to use filters correctly ?", because I understand the documentation and how it works, but how can I use it properly in mapstore code ?


Thank you very much for these clarifications (@offtherailz).

File: web/client/selectors/featuregrid.js (hunk 1 above)
Comment (Gaetanbrl) [reply web/client/selectors/featuregrid.js:284]: I will use this filter instead of spatialFilter : 

```
    {
            "format": "cql",
            "version": "1.0.0",
            "body": `${restrictedAreaOperatorSelector(state)}(${attribute}, ${asWKT})`,
            "id": `[${uniqueId("cql_restricted_area_")}]`,
            "restrictedArea": true
    };
```

**Comment by Gaetanbrl** [comment]
Here is a summary of the covered items.

## Changes Following Reviews

- Fixed the *maximum call stack* error.
- Applied **restricted area** logic only in **EDIT** mode.
- **Restricted area** is **never** applied for **ADMIN** users.
- **Restricted area** is applied only for **USER** and **ANONYMOUS** profiles.
- Added option to enable synchronization between the attribute table and the map (see [this comment](https://github.com/geosolutions-it/MapStore2/pull/10515#discussion_r1808831879)).
- Various technical feedback items addressed (@MV88).

## Outstanding Points

- Unit tests have not been implemented yet.
- There is an issue when synchronization is enabled and then the user switches to edit mode — the CQL syntax becomes invalid  

 This last error seems related to `srid` and CQL syntax. Refer to the [google group dev](https://groups.google.com/g/mapstore-developers/c/lJFzee_0oBI) ticket for more details.

**Comment by tdipisa** [comment]
@Gaetanbrl I'm sorry, thank you for this but this PR is quite old now. Do you have plans to re-handle this according to our review above and pending outstanding stuff? Otherwise, I would like to close the PR.

**Comment by Gaetanbrl** [comment]
Same as #10524.

I can look into updating the PR, but I’m not certain it will still be sufficient...

**Comment by tdipisa** [comment]
Hi @Gaetanbrl as you told me during our exchange last week, I'm going to close this.
